### PR TITLE
Update integration test framework

### PR DESCRIPTION
**Pull request by drognanar**
Allow to open solutions
Allow to save solutions
Add workspace operations (add reference, open document)
Create proxy classes for remote static methods
Make more RemotingHelper static methods internal rather than private
Allow to start integration test framework from a different hive
Add external visibility of the integration framework to test impact

### CODE REVIEW

File: src/VisualStudio/TestUtilities/Remoting/RemotingHelper.cs [hunk 1]
```diff
@@ -80,7 +79,19 @@ private static IWpfTextViewHost GetActiveTextViewHost()
...
+
+        #region Application Operations
+
+        public static void ExecuteOnUIThread(Action action)
+            => Application.Current.Dispatcher.Invoke(action);
+
+        public static T ExecuteOnUIThread<T>(Func<T> action)
+            => Application.Current.Dispatcher.Invoke(action);
+
+        internal static void WaitForApplicationIdle()
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/RemotingHelper.cs]: Make this public as well.

File: src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs [hunk 2]
```diff
@@ -20,24 +20,30 @@ public sealed class VisualStudioInstanceFactory : IDisposable
         internal static readonly string VsCommon7Folder = Path.GetFullPath(IntegrationHelper.GetRegistryKeyValue(Registry.LocalMachine, VsRegistryRoot, "InstallDir").ToString());
 
         internal static readonly string VsExeFile = Path.Combine(VsCommon7Folder, "devenv.exe");
-        internal static readonly string VsLaunchArgs = $"{(string.IsNullOrWhiteSpace(Settings.Default.VsRootSuffix) ? "/log" : $"/rootsuffix {Settings.Default.VsRootSuffix}")} /log";
+
+        internal string VsLaunchArgs(string rootSuffix) => $"{(string.IsNullOrWhiteSpace(rootSuffix) ? "/log" : $"/rootsuffix {rootSuffix}")} /log";
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs]: Please change back. The app.config file defines the `RootSuffix` setting, which can be changed in the TestImpact repo when the tests are run.

File: src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs [hunk 3]
```diff
@@ -20,24 +20,30 @@ public sealed class VisualStudioInstanceFactory : IDisposable
...
 
+        public VisualStudioInstanceFactory()
+        {
+        }
+
         /// <summary>
         /// Returns a <see cref="VisualStudioInstanceContext"/>, starting a new instance of Visual Studio if necessary.
         /// </summary>
-        public VisualStudioInstanceContext GetNewOrUsedInstance()
+        public VisualStudioInstanceContext GetNewOrUsedInstance(string rootSuffix = null)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs]: No. The rootSuffix is configured through the application settings.

File: src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs [hunk 4]
```diff
@@ -66,29 +72,29 @@ private void ThrowExceptionIfAlreadyHasActiveContext()
         /// <summary>
         /// Starts up a new <see cref="VisualStudioInstance"/>, shutting down any instances that are already running.
         /// </summary>
-        private void StartNewInstance()
+        private void StartNewInstance(string rootSuffix)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs]: Same as above.

File: src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs [hunk 5]
```diff
@@ -66,29 +72,29 @@ private void ThrowExceptionIfAlreadyHasActiveContext()
...
+            var process = StartNewVisualStudioProcess(rootSuffix);
 
             // We wait until the DTE instance is up before we're good
             var dte = IntegrationHelper.WaitForNotNullAsync(() => IntegrationHelper.TryLocateDteForProcess(process)).Result;
 
             _currentlyRunningInstance = new VisualStudioInstance(process, dte);
         }
 
-        private static Process StartNewVisualStudioProcess()
+        private Process StartNewVisualStudioProcess(string rootSuffix)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs]: Same as above.

File: src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs [hunk 6]
```diff
@@ -20,24 +20,30 @@ public sealed class VisualStudioInstanceFactory : IDisposable
...
+
+        internal string VsLaunchArgs(string rootSuffix) => $"{(string.IsNullOrWhiteSpace(rootSuffix) ? "/log" : $"/rootsuffix {rootSuffix}")} /log";
 
         /// <summary>
         /// The instance that has already been launched by this factory and can be reused.
         /// </summary>
         private VisualStudioInstance _currentlyRunningInstance;
         private bool _hasCurrentlyActiveContext;
 
+        public VisualStudioInstanceFactory()
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs]: Why do you need a public constructor?

File: src/VisualStudio/TestUtilities/Workspace/Solution.cs [hunk 7]
```diff
@@ -45,13 +45,27 @@ public string FileName
...
+            var dteProject = _dteSolution.AddFromFile(projectFilePath);
+            Save();
+            return new Project(dteProject, this, ProjectLanguage.CSharp);
+        }
+
         public Project AddProject(string projectName, ProjectTemplate projectTemplate, ProjectLanguage projectLanguage)
         {
             var solutionFolder = Path.GetDirectoryName(FileName);
-            var projectPath = Path.Combine(solutionFolder, projectName);
+            var projectPath = Path.Combine(solutionFolder, Path.GetFileNameWithoutExtension(projectName));
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Workspace/Solution.cs]: Why? A filename shouldn't be getting passed in here.

File: src/VisualStudio/TestUtilities/Workspace/Solution.cs [hunk 8]
```diff
@@ -45,13 +45,27 @@ public string FileName
...
             var projectTemplatePath = GetProjectTemplatePath(projectTemplate, projectLanguage);
 
             var dteProject = _dteSolution.AddFromTemplate(projectTemplatePath, projectPath, projectName, Exclusive: false);
+            Save();
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Workspace/Solution.cs]: Save should be explicit, not implicit.

File: src/VisualStudio/TestUtilities/Workspace/Solution.cs [hunk 9]
```diff
@@ -45,13 +45,27 @@ public string FileName
             }
         }
 
+        public void Save()
+        {
+            Directory.CreateDirectory(Path.GetDirectoryName(FileName));
+            DteSolution.SaveAs(FileName);
+        }
+
+        public Project OpenProject(string projectFilePath)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Workspace/Solution.cs]: AddExistingProject? Either that or this should be modified with a second parameter that indicates whether the project should be added to the existing solution or added to a new solution.

File: src/VisualStudio/TestUtilities/Proxy/RemotingProxy.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Reflection;
+using Roslyn.VisualStudio.Test.Utilities.Remoting;
+
+namespace Roslyn.VisualStudio.Test.Utilities.Proxy
+{
+    internal static class RemotingProxy
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Proxy/RemotingProxy.cs]: Why were these moved here? They were (and should be) only called from a single location.

File: src/VisualStudio/TestUtilities/Proxy/WorkspaceProxy.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Reflection;
+using Roslyn.VisualStudio.Test.Utilities.Remoting;
+
+namespace Roslyn.VisualStudio.Test.Utilities.Proxy
+{
+    internal class WorkspaceProxy
+    {
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Proxy/WorkspaceProxy.cs]: As above, I feel these should only be called in one location and should be inlined there.

File: src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.Shell;
+using RoslynProject = Microsoft.CodeAnalysis.Project;
+using RoslynWorkspace = Microsoft.CodeAnalysis.Workspace;
+
+namespace Roslyn.VisualStudio.Test.Utilities.Remoting
+{
+    internal class RemotingWorkspace
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs]: Rename to `WorkspaceWrapper` so that it matches the existing naming convention.

File: src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+            RoslynProject project = GetProject(workspace, projectName);
+            DocumentId documentId = project.Documents.FirstOrDefault(doc => doc.Name.Equals(documentName)).Id;
+
+            if (documentId != null)
+            {
+                RemotingHelper.ExecuteOnUIThread(() => workspace.OpenDocument(documentId, activate: true));
+            }
+        }
+
+        private static RoslynWorkspace GetWorkspace()
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs]: Rename to `CreateWorkspaceWrapper` and move to `RemotingHelper` to match the existing conventions.

File: src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        public static void AddReference(string projectName, string referenceAssemblyPath)
+        {
+            var metadataReference = MetadataReference.CreateFromFile(referenceAssemblyPath);
+            RoslynWorkspace workspace = GetWorkspace();
+            RoslynProject newProject = GetProject(workspace, projectName).AddMetadataReference(metadataReference);
+            workspace.TryApplyChanges(newProject.Solution);
+        }
+
+        public static void OpenDocument(string projectName, string documentName)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs]: This should be an instance method and `workspace` should be a field that is wrapped by the `WorkspaceWrapper` proxy.

File: src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+using Microsoft.VisualStudio.Shell;
+using RoslynProject = Microsoft.CodeAnalysis.Project;
+using RoslynWorkspace = Microsoft.CodeAnalysis.Workspace;
+
+namespace Roslyn.VisualStudio.Test.Utilities.Remoting
+{
+    internal class RemotingWorkspace
+    {
+        public static void AddReference(string projectName, string referenceAssemblyPath)
+        {
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs]: This should be an instance method and workspace should be a field that is wrapped by the WorkspaceWrapper proxy.

**Comment by tannergooding** [comment]
Done initial pass.

File: src/VisualStudio/TestUtilities/Workspace/Solution.cs (hunk 7 above)
Comment (drognanar) [inline src/VisualStudio/TestUtilities/Workspace/Solution.cs]: According to DTE solution AddFromTemplate docs:

> ProjectName: Required. The name of the project file in the destination directory. This should include the extension. The name displayed is derived from ProjectName.

Unless I misunderstood I believe that this means that the projectName should also contain the csproj ending.

File: src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs (hunk 13 above)
Comment (drognanar) [inline src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs]: I actually do not completely follow what convention is being set here.
RemotingWorkspace provides static methods to operate with the workspace inside of VS.
GetWorkspace is the static method that gets the workspace inside of VS.
WorkspaceProxy provides static methods to operate with the workspace outside of VS.

This is analogous to the way that:
RemotingHelper provides static methods to operate with the text editor inside of VS
GetActiveTextViewHost is the static method that gets the active text view host inside of VS.
EditorWindow provides methods to operate with the text editor outside of VS.

What is the reason that makes it necessary for `EditorWindow` to use static methods while while `Workspace` should be a `MarshalByRef` object?

Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/RemotingWorkspace.cs]: For most items (especially things such as the `Workspace` or `InteractiveWindow`, which represent a service), the convention should be to create a wrapper class around the object and then have a static method in `RemotingHelper` which returns an instance of said wrapper.

`GetActiveTextViewHost` is somewhat special because it is not a service and two sequential calls can potentially return different interfaces. It should (therefore) not be marshalled across the process boundary as it can lead to some hard-to-track down bugs (and can mess with the object lifetime in ways VS doesn't expect).

File: src/VisualStudio/TestUtilities/VisualStudioInstance.cs [hunk 16]
```diff
@@ -79,6 +80,14 @@ public VisualStudioInstance(Process process, DTE dte)
         internal T ExecuteOnHostProcess<T>(Type type, string methodName, BindingFlags bindingFlags, params object[] parameters)
              => ExecuteOnHostProcess<T>(type.Assembly.Location, type.FullName, methodName, bindingFlags, parameters);
 
+        internal void ExecuteOnHostProcess(Type type, string methodName, BindingFlags bindingFlags, params object[] parameters)
+             => ExecuteOnHostProcess(type.Assembly.Location, type.FullName, methodName, bindingFlags, parameters);
+
+        internal void ExecuteOnHostProcess(string assemblyFilePath, string typeFullName, string methodName, BindingFlags bindingFlags, params object[] parameters)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/VisualStudioInstance.cs]: Why?

File: src/VisualStudio/TestUtilities/Proxy/VisualStudioApplication.cs [hunk 17]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Reflection;
+using Roslyn.VisualStudio.Test.Utilities.Remoting;
+
+namespace Roslyn.VisualStudio.Test.Utilities.Proxy
+{
+    internal class VisualStudioApplication
+    {
+        public static void WaitForApplicationIdle(VisualStudioInstance host)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Proxy/VisualStudioApplication.cs]: If anything, this method should be a public member of the `VisualStudioInstance` class.

File: src/VisualStudio/TestUtilities/Workspace/Solution.cs (hunk 7 above)
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Workspace/Solution.cs]: It would likely provide a better experience by appending the extension below based on the project language we are given (I believe it would also match the current TAO behavior, which is desired where possible).

**Comment by drognanar** [comment]
FYI @jasonmalinowski @dotnet/testimpact

File: src/VisualStudio/TestUtilities/Remoting/WorkspaceWrapper.cs [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    internal class WorkspaceWrapper : MarshalByRefObject
+    {
+        private RoslynWorkspace _workspace;
+
+        public WorkspaceWrapper(RoslynWorkspace workspace)
+        {
+            _workspace = workspace;
+        }
+
+        public void AddReference(string projectName, string referenceAssemblyPath)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/WorkspaceWrapper.cs]: An exception will get thrown if `MetadataReference.CreateFromFile` or `RoslynProject.AddMetadataReference` fails, but how do we let the caller know if `TryApplyChanges` doesn't succeed?

File: src/VisualStudio/TestUtilities/Remoting/WorkspaceWrapper.cs [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+            RoslynProject newProject = GetProject(projectName).AddMetadataReference(metadataReference);
+            _workspace.TryApplyChanges(newProject.Solution);
+        }
+
+        public void OpenDocument(string projectName, string documentName)
+        {
+            RoslynProject project = GetProject(projectName);
+            DocumentId documentId = project.Documents.FirstOrDefault(doc => doc.Name.Equals(documentName)).Id;
+
+            if (documentId != null)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/WorkspaceWrapper.cs]: I think we should throw an exception if `documentId == null`, otherwise the call to `OpenDocument` could do nothing and the caller could end up modifying an unexpected document.

File: src/VisualStudio/TestUtilities/VisualStudioTestUtilities.csproj [hunk 20]
```diff
@@ -42,6 +43,7 @@
     <Compile Include="Workspace\SolutionExplorer.cs" />
   </ItemGroup>
   <ItemGroup>
+    <InternalsVisibleToTestImpact Include="Microsoft.CodeAnalysis.TestImpact.VisualStudio.IntegrationTests" />
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/VisualStudioTestUtilities.csproj]: I think this just needs to be 'InternalsVisibleTo' now?

File: src/VisualStudio/TestUtilities/Remoting/WorkspaceWrapper.cs [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using RoslynProject = Microsoft.CodeAnalysis.Project;
+using RoslynWorkspace = Microsoft.CodeAnalysis.Workspace;
+
+namespace Roslyn.VisualStudio.Test.Utilities.Remoting
+{
+    internal class WorkspaceWrapper : MarshalByRefObject
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/WorkspaceWrapper.cs]: Where is `WorkspaceWrapper` being exposed?

File: src/VisualStudio/TestUtilities/Remoting/RemotingHelper.cs [hunk 22]
```diff
@@ -28,6 +29,14 @@ public static InteractiveWindowWrapper CreateCSharpInteractiveWindowWrapper()
             return new InteractiveWindowWrapper(vsInteractiveWindow.InteractiveWindow);
         }
 
+        private static WorkspaceWrapper CreateWorkspaceWrapper()
+        {
+            var componentModel = (IComponentModel)(ServiceProvider.GlobalProvider.GetService(typeof(SComponentModel)));
```
Comment (jasonmalinowski) [inline src/VisualStudio/TestUtilities/Remoting/RemotingHelper.cs]: Avoid using ServiceProvider.GlobalProvider if you can. Do we have a better context for this object we can use? Do we already know if the package has been loaded?

File: src/VisualStudio/TestUtilities/VisualStudioInstance.cs [hunk 23]
```diff
@@ -79,6 +79,16 @@ public VisualStudioInstance(Process process, DTE dte)
...
+        {
+            var objectUri = _service.Execute(assemblyFilePath, typeFullName, methodName, bindingFlags, parameters);
+
+            Debug.Assert(string.IsNullOrWhiteSpace(objectUri), "Expects to call this method on void methods.");
```
Comment (jasonmalinowski) [inline src/VisualStudio/TestUtilities/VisualStudioInstance.cs]: Don't assert, throw, since this means something is actually wrong.

File: src/VisualStudio/TestUtilities/VisualStudioInstance.cs [hunk 24]
```diff
@@ -136,6 +149,7 @@ public void Close()
 
         public void CloseAndDeleteOpenSolution()
         {
+            WaitForApplicationIdle();
```
Comment (jasonmalinowski) [inline src/VisualStudio/TestUtilities/VisualStudioInstance.cs]: Why? If there's *every* a wait there should always be a clear explanation as to why it's necessary in a comment. Random waits like this is the beginning of the end of any integration test framework. (I'd like that not to start now. :-))

File: src/VisualStudio/TestUtilities/VisualStudioInstance.cs [hunk 25]
```diff
@@ -122,6 +132,9 @@ internal Task<Window> LocateDteWindowAsync(string windowTitle)
         internal Task WaitForDteCommandAvailabilityAsync(string command)
             => IntegrationHelper.WaitForResultAsync(() => IntegrationHelper.RetryDteCall(() => Dte.Commands.Item(command).IsAvailable), expectedResult: true);
 
+        public void WaitForApplicationIdle()
```
Comment (jasonmalinowski) [inline src/VisualStudio/TestUtilities/VisualStudioInstance.cs]: I do not think there should be a wait API like this at this layer. All waits should live in code that's running in VS, as otherwise it'll be unclear where the waits need to happen.

File: src/VisualStudio/TestUtilities/Remoting/RemotingHelper.cs [hunk 26]
```diff
@@ -80,7 +83,19 @@ private static IWpfTextViewHost GetActiveTextViewHost()
...
+
+        #region Application Operations
+
+        public static void ExecuteOnUIThread(Action action)
+            => Application.Current.Dispatcher.Invoke(action);
+
+        public static T ExecuteOnUIThread<T>(Func<T> action)
+            => Application.Current.Dispatcher.Invoke(action);
+
+        public static void WaitForApplicationIdle()
```
Comment (jasonmalinowski) [inline src/VisualStudio/TestUtilities/Remoting/RemotingHelper.cs]: Mentioned elsewhere: let's not have this. This is a leaky abstraction.

**Comment by drognanar** [comment]
I have rebased the change on top of master and addressed the feedback.
The PR was simplified as PR #11223 incorporated a few of the same features.

File: src/VisualStudio/TestUtilities/Workspace/Project.cs [hunk 27]
```diff
@@ -35,5 +35,7 @@ internal Project(DteProject dteProject, Solution solution, ProjectLanguage langu
         public ProjectLanguage Language => _language;
 
         public Solution Solution => _solution;
+
+        public void OpenFile(string fileName) => DteProject.DTE.ExecuteCommandAsync("File.OpenFile", fileName).GetAwaiter().GetResult();
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Workspace/Project.cs]: Make `OpenFileAsync`

File: src/VisualStudio/TestUtilities/Remoting/RemotingHelper.cs [hunk 28]
```diff
@@ -52,6 +52,11 @@ public static string ActiveTextViewContents
             }
         }
 
+        public static IWpfTextViewMargin GetTextViewMargin(string marginName)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Remoting/RemotingHelper.cs]: Inline

File: src/VisualStudio/TestUtilities/Workspace/Solution.cs [hunk 29]
```diff
@@ -68,20 +68,33 @@ public Project AddProject(string projectName, ProjectTemplate projectTemplate, P
...
+                dteProject = GetDteProject(projectName);
+            }
 
-                foreach (DteProject project in dteSolutionProjects)
-                {
-                    var dteProjectName = IntegrationHelper.RetryRpcCall(() => project.Name);
+            return new Project(dteProject, this, projectLanguage);
+        }
+
+        public Project GetOpenedProject(string projectName, ProjectLanguage projectLanguage)
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Workspace/Solution.cs]: `GetProject`?

File: src/VisualStudio/TestUtilities/Workspace/Solution.cs [hunk 30]
```diff
@@ -68,20 +68,33 @@ public Project AddProject(string projectName, ProjectTemplate projectTemplate, P
...
+                var dteProjectName = IntegrationHelper.RetryRpcCall(() => project.Name);
+
+                if (dteProjectName == projectName)
+                {
+                    return project;
                 }
             }
 
-            return new Project(dteProject, this, projectLanguage);
+            return null;
```
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/Workspace/Solution.cs]: throw an Exception instead?

**Comment by tannergooding** [comment]
LGTM, modulo the new comment.

**Comment by jasonmalinowski** [comment]
:+1:
